backupccl: fix flake in backup stats tests

pbardea opened the pull request:
This commit updates the way we generate table stats in tests to ensure
that the stats cache always sees the values we insert. This change takes
advantage of the fact that inject stats directly to a table will
invalidate the stats cache, so when BACKUP queries the stats it will
find the latest information.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52639)
<!-- Reviewable:end -->

dt commented on pkg/ccl/backupccl/backup_job.go:
```diff
@@ -350,7 +350,8 @@ func backup(
 		if tableDesc := sqlbase.TableFromDescriptor(&backupManifest.Descriptors[i], hlc.Timestamp{}); tableDesc != nil {
-			// Collect all the table stats for this table.
+			// Invalidate the stats for this table to ensure they are up to date.
+			statsCache.InvalidateTableStats(ctx, tableDesc.GetID())
```
Hm. 

Could this also mean other traffic sees a blip every time backup runs and evicts the cache?
Also, I guess compared to all the actual data exporting backup does blocking on re-reading stats shouldn't be _that_ bad, but do we have a sense for how bad it might be for e.g. 1k tables doing small/fast inc backups?

pbardea replied:
Ya, I also felt a bit uneasy. Although I don't think the perf hit would be too bad, it's a bit of an uneasy backport. I found that directly inject statistics invalidates the stats cache so instead I opted to just load stats like that to unflake the tests.

pbardea committed:
backupccl: fix flake in backup stats tests

This commit updates the way we generate table stats in tests to ensure
that the stats cache always sees the values we insert. This change takes
advantage of the fact that inject stats directly to a table will
invalidate the stats cache, so when BACKUP queries the stats it will
find the latest information.

Release note: None

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52639#-:-MF0PUSM-Gp92zF_r1vg:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt)



<!-- Sent from Reviewable.io -->

pbardea commented:
TFTRs!
bors r=dt,rytaft
